sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/FileIndexer.java

```diff
@@ -132,84 +141,95 @@ private static String pluralizeFiles(int count) {
+      new IndexFileVisitor(type, progress));
   }
 
-  private Void indexFile(DefaultModuleFileSystem fileSystem, Progress progress, Path sourceFile, InputFile.Type type) throws IOException {
+  private Void indexFile(Path sourceFile, InputFile.Type type, Progress progress) throws IOException {
     // get case of real file without resolving link
-    Path realFile = sourceFile.toRealPath(LinkOption.NOFOLLOW_LINKS);
-    DefaultInputFile inputFile = inputFileBuilder.create(realFile, type, fileSystem.encoding());
-    if (inputFile != null) {
-      if (exclusionFilters.accept(inputFile, type) && accept(inputFile)) {
-        String parentRelativePath = getParentRelativePath(fileSystem, inputFile);
-        synchronized (this) {
-          fileSystem.add(inputFile);
-          indexParentDir(fileSystem, inputFile, parentRelativePath);
-          progress.markAsIndexed(inputFile);
-        }
-        LOG.debug("'{}' indexed {}with language '{}'", inputFile.relativePath(), type == Type.TEST ? "as test " : "", inputFile.language());
-        inputFileBuilder.checkMetadata(inputFile);
-      } else {
-        progress.increaseExcludedByPatternsCount();
-      }
+    Path realAbsoluteFile = sourceFile.toRealPath(LinkOption.NOFOLLOW_LINKS).toAbsolutePath().normalize();
+    String relativePathStr = PathResolver.relativePath(module.getBaseDir(), realAbsoluteFile);
+    if (relativePathStr == null) {
+      LOG.warn("File '{}' is ignored. It is not located in module basedir '{}'.", realAbsoluteFile.toAbsolutePath(), module.getBaseDir());
+      return null;
+    }
+    Path relativePath = module.getBaseDir().relativize(realAbsoluteFile);
+    if (!exclusionFilters.accept(realAbsoluteFile, relativePath, type)) {
+      progress.increaseExcludedByPatternsCount();
+      return null;
+    }
+    String language = langDetection.language(realAbsoluteFile, relativePath);
+    if (language == null && langDetection.forcedLanguage() != null) {
+      LOG.warn("File '{}' is ignored because it doesn't belong to the forced language '{}'", realAbsoluteFile.toAbsolutePath(), langDetection.forcedLanguage());
+      return null;
+    }
+    DefaultInputFile inputFile = inputFileBuilder.create(type, relativePathStr, language);
+    if (!accept(inputFile)) {
+      progress.increaseExcludedByPatternsCount();
+      return null;
+    }
+    String parentRelativePath = getParentRelativePath(inputFile);
+    synchronized (this) {
+      indexParentDir(inputFile, parentRelativePath);
+      progress.markAsIndexed(inputFile);
     }
+    LOG.debug("'{}' indexed {}with language '{}'", relativePathStr, type == Type.TEST ? "as test " : "", inputFile.language());
+    inputFileBuilder.checkMetadata(inputFile);
     return null;
   }
 
-  private static String getParentRelativePath(DefaultModuleFileSystem fileSystem, InputFile inputFile) {
+  private String getParentRelativePath(InputFile inputFile) {
     Path parentDir = inputFile.path().getParent();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "path"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)